internal/server/server.go

```diff
@@ -0,0 +1,240 @@
+	"github.com/gin-gonic/gin"
+	"log"
+	"net/http"
+	"time"
+)
+
+// Env defines routes, as well as objects that should be shared across requests
+// (database connections, config, etc.)
+//
+// Defining routes like this allows for easy dependency injection in tests.
+type Env struct {
+	db      database.Database
+	webroot string
+}
+
+func NewEnv(db database.Database, webroot string) *Env {
+	return &Env{db, webroot}
+}
+
+func (env *Env) ping(c *gin.Context) {
+	c.String(http.StatusOK, "pong")
+}
+
+func (env *Env) serveIndex(c *gin.Context) {
+	now := time.Now()
+
+	csrfToken, err := crypto.GenerateToken()
+	if err != nil {
+		log.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	session, err := env.db.CreatePreAuthSession(csrfToken, now)
+	if err != nil {
+		log.Println(err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	c.Header("Content-Type", "text/html")
+	sessionCookie := http.Cookie{
+		Name:     "session_token",
+		Value:    session.SessionToken,
+		Path:     "/",
+		Secure:   true,
+		HttpOnly: true,
+		SameSite: http.SameSiteLaxMode,
+	}
+
+	// Gin has a slightly weirder interface for cookies for some reason
+	http.SetCookie(c.Writer, &sessionCookie)
+	c.HTML(http.StatusOK, "index.html", gin.H{"csrf": csrfToken})
+}
+
+func (env *Env) login(c *gin.Context) {
+	now := time.Now()
+	csrfToken := c.Request.Header.Get("CSRF")
+
+	sessionToken, err := c.Cookie("session_token")
+	// should look into deduplicating this
+	// should also look into a logging library w/ levels (debug, info, warning, etc)
+	if err != nil {
+		log.Println("Missing session token")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	session, err := env.db.FetchSession(sessionToken)
+	if err != nil {
+		log.Printf("Error fetching session: %v\n", err)
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if session.Expired(now) {
+		log.Println("Session expired")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if !crypto.IsSessionOwner(session, csrfToken) {
+		log.Println("Not owner of session")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	if session.Authenticated() {
+		log.Println("Already logged in")
+		c.AbortWithStatus(http.StatusOK)
+		return
+	}
+
+	email, password, ok := c.Request.BasicAuth()
+	if !ok {
+		log.Println("Missing basic auth")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	account, err := env.db.FetchAccount(email)
+	if err != nil {
+		log.Println("Missing account")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
+
+	isCorrectPassword, err := crypto.IsCorrectPassword(account, password)
+	if err != nil {
+		log.Println("Salt stored in database is incorrectly encoded")
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	if !isCorrectPassword {
+		log.Println("Bad password")
+		c.AbortWithStatus(http.StatusUnauthorized)
+		return
+	}
```

Comment:
I don't think you need to split this out, just consolidate both checks and into `crypto.IsCorrectPassword` and return an error if either one fails.